src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,576 @@
+            return await backupManager.requestKeyBackupVersion();
+        },
+
+        getCryptoEventEmitter(): TypedEventEmitter<RustBackupCryptoEvents, RustBackupCryptoEventMap> {
+            return rustCrypto;
+        },
+    };
+}
+
+/**
+ * Enumerates the different kind of errors that can occurs when downloading and importing a key from backup.
+ */
+export enum KeyDownloadError {
+    /** The backup version in use is out of sync with the server version. */
+    VERSION_MISMATCH = "VERSION_MISMATCH",
+    /** The requested key is not in the backup. */
+    MISSING_DECRYPTION_KEY = "MISSING_DECRYPTION_KEY",
+    /** A network error occurred while trying to get the key. */
+    NETWORK_ERROR = "NETWORK_ERROR",
+    /** The loop as been stopped. */
+    STOPPED = "STOPPED",
+    /** An unknown error occurred while decrypting/importing the key */
+    UNKNOWN_ERROR = "UNKNOWN_ERROR",
+    /** The server is rate limiting us. */
+    RATE_LIMITED = "RATE_LIMITED",
+    /** The backup is not configured correctly, can be that there is no backup, that it's not trusted
+     * , that we don't have the correct key in cache... */
+    CONFIGURATION_ERROR = "CONFIGURATION_ERROR",
+}
+
+/** Helper type for requested session*/
+type SessionInfo = { roomId: string; megolmSessionId: string };
+
+/** Helper type for the result of a key download. */
+type KeyDownloadResult =
+    | { ok: true; value: KeyBackupSession }
+    | { ok: false; error: KeyDownloadError; [key: string]: any };
+
+/** Holds the current backup decryptor and version that should be used. */
+type Configuration = {
+    backupVersion: string;
+    decryptor: RustBackupDecryptor;
+};
+
+/**
+ * Signaling for the Downloader loop.
+ * Not yet used by API, yet useful for testing.
+ */
+export enum KeyDownloaderEvent {
+    DownloadLoopStateUpdate = "download_loop_started",
+    DownLoopStep = "download_loop_step",
+    QueryKeyError = "query_key_error",
+    KeyImported = "key_imported",
+}
+export type KeyDownloaderEventMap = {
+    [KeyDownloaderEvent.DownloadLoopStateUpdate]: (loopRunning: boolean) => void;
+    [KeyDownloaderEvent.DownLoopStep]: (remaining: number) => void;
+    [KeyDownloaderEvent.QueryKeyError]: (errCode: KeyDownloadError) => void;
+    [KeyDownloaderEvent.KeyImported]: (roomId: string, sessionId: string) => void;
+};
+
+/**
+ * This function is called when an 'unable to decrypt' error occurs. It attempts to download the key from the backup.
+ *
+ * The current backup API lacks pagination, which can lead to lengthy key retrieval times for large histories (several 10s of minutes).
+ * To mitigate this, keys are downloaded on demand as decryption errors occurs.
+ * While this approach may result in numerous requests, it improves user experience by reducing wait times for message decryption.
+ * 
+ * The PerSessionKeyBackupDownloader is resistant to backup configuration changes, it will automatically resume querying when
+ * the backup is configured correctly.
```

Comment:
```suggestion
 * The PerSessionKeyBackupDownloader is resistant to backup configuration changes: it will automatically resume querying when
 * the backup is configured correctly.
```